README.md
[![Build Status](https://travis-ci.org/kadena-io/merkle-log.svg?branch=master)](https://travis-ci.org/kadena-io/merkle-log)

# Haskell Implementation of Merkle Tree Logs

This packages implements Merkle Tree Logs similar to those described in RFC 6962
in Haskell.

Merkle Logs are a append-only data structure. The tree layout in this
implementation of Merkle trees is based on the description of Merkle trees in
RFC 6962. With this tree layout extending a Merkle tree requires chaining a
logarithmic number of nodes at the end of the tree. Unlike RFC 6962 the Merkle
trees in this module support the creation of unbalanced MerkleTrees by nesting
sub-trees as leafs of Merkle trees. Also, unlike RFC 6962 this module generates
fully self-contained inclusion proofs that don't rely on the client being aware
of the balancing of the Merkle Tree that was used to generate the proof.

## Format

The implementation stores Merkle trees in a packed format in memory. This allows
for efficient construction, serialization, and querying. Trees are opaque
objects that are allocated and deallocated as well as serialized and
deserialized as a whole, which matches many use cases. Also, trees can be
nested, by building larger Merkle trees that have smaller trees as inputs to
their leafs.

The overhead per indexed item is 64 bytes when 256 bit hashes are used. Thus,
about 16,000 items can be index in 1MB of memory.

We plan to make the trees extensible and support loading and storing trees in
chunks that represent immutable full subtrees. Please file an issue on GitHub if
you need this feature.

## Proofs

Proofs are self contained and don't rely on a particular implementation of
Merkle tree. In particular, proofs don't depend on how the tree is balanced.

A proof contains the proof subject (the input for which inclusion is proven) as
a plain `ByteString`. The result of validating a proof is a Merkle tree root
that must match the root of the Merkle tree that includes the subject. A proof
doesn't include the root hash of the Merkle tree, because the root must be
obtained from a trusted / authenticated source. Including it in the proof would
thus be redundant and may even be misleading.

At the moment only inclusion / audit proofs are supported. We plan to also
implement consistency proofs. Please file an issue on GitHub if you need
consistency proofs.

## Build and Installation

The package can be build with cabal via

```sh
cabal new-update
cabal new-build merkle-log
```

The test suite can be run with

```sh
cabal new-test merkle-log
```

Benchmarks are available via


```sh
cabal new-bench merkle-log
```

## Benchmarks

The following benchmark results compare the performance of this package with
the Merkle tree implementations in the packages
[merkle-tree](http://hackage.haskell.org/package/merkle-tree) and
[hash-tree](http://hackage.haskell.org/package/hash-tree) for different hash
functions on a 64bit Intel CPU.

Create tree:

![Create Tree Results](bench/results/create-tree.png)

Create inclusion proof:

![Create Inclusion Proof Results](bench/results/create-inclusion-proof.png)

Verify inclusion proof:

![Verify Inclusion Proof Results](bench/results/verify-inclusion-proof.png)



CHANGELOG.md
# Revision history for merkle-log

## 0.2.0 -- 2023-06-08

*   Switch from cryptonite to crypton package.

## 0.1.1 -- 2021-06-24

*   Build with GHC-9

## 0.1.0.0 -- 2019-05-28

*   First version. Released on an unsuspecting world.


test/Main.hs
{-# LANGUAGE CPP #-}
{-# LANGUAGE FlexibleContexts #-}
{-# LANGUAGE FlexibleInstances #-}
{-# LANGUAGE LambdaCase #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE TypeApplications #-}

{-# OPTIONS_GHC -fno-warn-orphans #-}

-- |
-- Module: Main
-- Copyright: Copyright © 2019 Kadena LLC.
-- License: MIT
-- Maintainer: Lars Kuhtz <lars@kadena.io>
-- Stability: experimental
--
-- TODO
--
module Main
( main

-- * Properties
, properties
, prop_proof
, prop_proofExhaustive
, prop_proofSize
, prop_encodeProofObject
, prop_encodeMerkleRoot
, prop_encodeMerkleTree
) where

import Control.DeepSeq
import Control.Monad.Catch

import Crypto.Hash.Algorithms (SHA512t_256, HashAlgorithm)

import Data.Bitraversable
import qualified Data.ByteArray as BA
import qualified Data.ByteString as B
import qualified Data.ByteString.Char8 as B8
import qualified Data.List.NonEmpty as NE

import System.Exit

import Test.QuickCheck

-- internal modules

import Data.MerkleLog

-- -------------------------------------------------------------------------- --
-- Support for QuickCheck < 2.12

#if ! MIN_VERSION_QuickCheck(2,12,0)
infix 4 =/=
(=/=) :: (Eq a, Show a) => a -> a -> Property
x =/= y = counterexample (show x ++ interpret res ++ show y) res
  where
    res = x /= y
    interpret True  = " /= "
    interpret False = " == "

isSuccess :: Result -> Bool
isSuccess Success{} = True
isSuccess _ = False
#endif

-- -------------------------------------------------------------------------- --
-- Main

main :: IO ()
main = do
    results <- traverse (bitraverse print quickCheckResult) properties
    if and $ isSuccess . snd <$> results
        then exitSuccess
        else exitFailure

-- | Properties
--
properties :: [(String, Property)]
properties =
    [ ("create merkle tree and confirm the size", property prop_tree)
    , ("create and verify merkle proof", property prop_proof)
    , ("create and verify merkle proof for all tree items for tree of size 30", prop_proofExhaustive 30)
    , ("create and verify merkle proof for tree of size 1000 with items of sizes up to 1000 bytes", prop_proofSize 1000 1000)
    , ("creating proof for invalid input fails", property prop_proofInvalidInput)
    , ("running proof with invalid subject fails", property prop_proofInvalidSubject)
    , ("running proof with invalid object path fails", property prop_proofInvalidObjectPath)
    , ("running proof with invalid object hash fails", property prop_proofInvalidObjectHash)
    , ("running proof with invalid object step count fails", property prop_proofInvalidStepCount)
    , ("create and verify merkle proof for nested trees", property prop_chainProof)
    , ("encoding roundtrip for merkle proof object", property prop_encodeProofObject)
    , ("encoding roundtrip for merkle proof chain object", property prop_encodeProofChainObject)
    , ("encoding roundtrip for merkle root", property prop_encodeMerkleRoot)
    , ("encoding roundtrip for merkle tree", property prop_encodeMerkleTree)
    ]

-- -------------------------------------------------------------------------- --
-- Utils

nodeCount :: Int -> Int
nodeCount i = max 1 (2 * i - 1)
{-# INLINE nodeCount #-}

-- | Change diretion of first proof step. Throws error is the proof
-- is empty (singleton tree).
--
changeProofPath :: HashAlgorithm a => MerkleProof a -> MerkleProof a
changeProofPath p = p { _merkleProofObject = obj }
  where
    obj = case decodeMerkleProofObject (BA.pack @BA.Bytes o) of
        Right x -> x
        Left e -> error $ "test/Main.chainProofPath: failed to decode proof object: " <> show e
    o = case splitAt 12 (BA.unpack (_merkleProofObject p)) of
        (h, 0x00 : t) -> h <> (0x01 : t)
        (h, 0x01 : t) -> h <> (0x00 : t)
        (_, _ : _) -> error "invalid proof object"
        (_, []) -> error "unexpected empty proof object"


-- | Change hash of first proof step. Throws error is the proof
-- is empty (singleton tree).
--
changeProofHash :: HashAlgorithm a => MerkleProof a -> MerkleProof a
changeProofHash p = p { _merkleProofObject = obj }
  where
    obj = case decodeMerkleProofObject (BA.pack @BA.Bytes o) of
        Right x -> x
        Left e -> error $ "test/Main.chainProofHash: failed to decode proof object: " <> show e
    o = case splitAt 12 (BA.unpack (_merkleProofObject p)) of
        (h, h1 : h2 : t) -> h <> (h1 : 1 + h2 : t)
        (_, []) -> error "unexpected empty proof object"
        _ -> error "invalid proof object"

-- | Changes the proof step count and verifies that decoding of the modified proof object fails.
-- Throws error is the proof is empty (singleton tree).
--
changeProofStepCount :: forall a . HashAlgorithm a => MerkleProof a -> Bool
changeProofStepCount p = case r of
    Left _ -> True
    Right _ -> False
  where
    r = decodeMerkleProofObject @a . BA.pack @BA.Bytes
        $ case splitAt 3 (BA.unpack (_merkleProofObject p)) of
            (h, c : t) -> h <> (c + 1 : t)
            (_, []) -> error "unexpected empty proof object"

-- -------------------------------------------------------------------------- --
-- Generators

newtype UniqueInputs a = UniqueInputs [MerkleNodeType a B.ByteString]
    deriving Show

instance HashAlgorithm a => Arbitrary (UniqueInputs a) where
    arbitrary = UniqueInputs
        . zipWith (\a () -> InputNode $ B8.pack (show a)) [0 :: Int .. ]
        <$> arbitrary

instance HashAlgorithm a => Arbitrary (MerkleNodeType a B.ByteString) where
    arbitrary = oneof
        [ InputNode . B.pack <$> arbitrary
        , TreeNode <$> arbitrary
        ]

instance HashAlgorithm a => Arbitrary (MerkleRoot a) where
    arbitrary = merkleNode <$> arbitrary <*> arbitrary

instance HashAlgorithm a => Arbitrary (MerkleHash a) where
    arbitrary = merkleLeaf @a . B.pack <$> arbitrary

instance HashAlgorithm a => Arbitrary (MerkleTree a) where
    arbitrary = merkleTree <$> arbitrary @[MerkleNodeType a B.ByteString]

instance HashAlgorithm a => Arbitrary (MerkleProof a) where
    arbitrary = go `suchThatMap` either (const Nothing) Just
      where
        go = do
            NonEmpty l <- arbitrary @(NonEmptyList (MerkleNodeType a B.ByteString))
            i <- choose (0, length l - 1)
            return (merkleProof (l !! i) i (merkleTree l))

-- | A chain of nested Merkle trees.
--
newtype MerkleTreeChain a = MerkleTreeChain
    { _getMerkleTreeChain :: NE.NonEmpty (Int, MerkleTree a)
        -- ^ a list of of merkle trees along with the position of the previous
        -- tree in the chain
    }
    deriving Show

genTrees
    :: forall a
    . HashAlgorithm a
    => Gen (MerkleTreeChain a)
genTrees = do
    a <- genTree (InputNode "a")
    i <- choose @Int (0, 10)
    MerkleTreeChain . (NE.:|) a <$> go i (merkleRoot $ snd a)
  where
    genTree x = do
        il <- arbitrary @[MerkleNodeType a B.ByteString]
        ir <- arbitrary
        return (length il , merkleTree (concat [il, pure x, ir]))

    go 0 _ = return []
    go i r = do
        a <- genTree (TreeNode r)
        (:) a <$> go (pred i) (merkleRoot $ snd a)

instance HashAlgorithm a => Arbitrary (MerkleTreeChain a) where
    arbitrary = genTrees

-- -------------------------------------------------------------------------- --
-- Properties

prop_tree :: [MerkleNodeType SHA512t_256 B.ByteString] -> Property
prop_tree l = size t === nodeCount (length l) .&. leafCount t === length l
  where
    t = force $ merkleTree @SHA512t_256 l

prop_proof :: [MerkleNodeType SHA512t_256 B.ByteString] -> NonNegative Int -> Property
prop_proof l (NonNegative i) = i < length l ==> runMerkleProof p === merkleRoot t
  where
    t = merkleTree @SHA512t_256 l
    p = case merkleProof (l !! i) i t of
        Left e -> error (displayException e)
        Right x -> x

-- | Runtime is quadradic in the input parameter. 50 ~ 1sec, 100 ~ 5sec.
--
prop_proofExhaustive :: Int -> Property
prop_proofExhaustive n = once $ conjoin
    [ prop_proof ((InputNode . B.singleton . fromIntegral) <$> [0 .. i]) (NonNegative j)
    | i <- [0..n]
    , j <- [0..i]
    ]

-- | Runtime of @testSize n m@ can be expected to be bounded by @Ω(n * m)@.
-- @testSize 1000 1000@ ~ 1sec.
--
prop_proofSize :: Int -> Int -> Property
prop_proofSize n m = once $ do
    l <- vectorOf n (resize m arbitrary)
    i <- choose (0, n - 1)
    return $ prop_proof l (NonNegative i)

prop_proofInvalidInput
    :: [MerkleNodeType SHA512t_256 B.ByteString]
    -> NonNegative Int
    -> Property
prop_proofInvalidInput a (NonNegative i) = i < length a && notElem (InputNode "a") a
    ==> case merkleProof (InputNode "a") i (merkleTree @SHA512t_256 a) of
        Left _ -> True
        Right _ -> False

prop_proofInvalidSubject
    :: [MerkleNodeType SHA512t_256 B.ByteString]
    -> NonNegative Int
    -> Property
prop_proofInvalidSubject l (NonNegative i) = i < length l
    ==> runMerkleProof p' =/= merkleRoot t
  where
    t = merkleTree @SHA512t_256 l
    p = case merkleProof (l !! i) i t of
        Left e -> error (displayException e)
        Right x -> x
    p' = p { _merkleProofSubject = changeSubject (_merkleProofSubject p) }
    changeSubject (MerkleProofSubject (InputNode "a")) = MerkleProofSubject (InputNode "b")
    changeSubject _ = MerkleProofSubject (InputNode "a")

prop_proofInvalidObjectPath
    :: UniqueInputs SHA512t_256
    -> NonNegative Int
    -> Property
prop_proofInvalidObjectPath (UniqueInputs l) (NonNegative i)
    = length l > 1 && i < length l
    ==> runMerkleProof (changeProofPath p) =/= merkleRoot t
  where
    t = merkleTree @SHA512t_256 l
    p = case merkleProof (l !! i) i t of
        Left e -> error (displayException e)
        Right x -> x

prop_proofInvalidStepCount
    :: NonEmptyList (MerkleNodeType SHA512t_256 B.ByteString)
    -> NonNegative Int
    -> Property
prop_proofInvalidStepCount (NonEmpty l) (NonNegative i)
    = i < length l ==> changeProofStepCount p
  where
    t = merkleTree @SHA512t_256 l
    p = case merkleProof (l !! i) i t of
        Left e -> error (displayException e)
        Right x -> x

prop_proofInvalidObjectHash
    :: NonEmptyList (MerkleNodeType SHA512t_256 B.ByteString)
    -> NonNegative Int
    -> Property
prop_proofInvalidObjectHash (NonEmpty l) (NonNegative i)
    = 1 < length l && i < length l
    ==> runMerkleProof (changeProofHash p) =/= merkleRoot t
  where
    t = merkleTree @SHA512t_256 l
    p = case merkleProof (l !! i) i t of
        Left e -> error (displayException e)
        Right x -> x

prop_chainProof :: MerkleTreeChain SHA512t_256 -> Property
prop_chainProof (MerkleTreeChain l)
    = runMerkleProof @SHA512t_256 p === merkleRoot (snd $ NE.last l)
  where
    p = case merkleProof_ (InputNode "a") l of
        Right x -> x
        Left e -> error $ "test/Main.prop_chainProof: merkleProof failed: " <> show e

prop_encodeProofObject :: MerkleProof SHA512t_256 -> Property
prop_encodeProofObject p
    = case decodeMerkleProofObject (encodeMerkleProofObject @BA.Bytes po) of
        Left e -> error (displayException e)
        Right x -> po === x
  where
    po = _merkleProofObject p

prop_encodeProofChainObject :: MerkleTreeChain SHA512t_256 -> Property
prop_encodeProofChainObject (MerkleTreeChain l)
    = case decodeMerkleProofObject (encodeMerkleProofObject @BA.Bytes po) of
        Left e -> error (displayException e)
        Right x -> po === x
  where
    p = case merkleProof_ (InputNode "a") l of
        Left e -> error (displayException e)
        Right x -> x
    po = _merkleProofObject p

prop_encodeMerkleRoot :: MerkleTree SHA512t_256 -> Property
prop_encodeMerkleRoot t
    = case decodeMerkleRoot (encodeMerkleRoot @BA.Bytes r) of
        Left e -> error (displayException e)
        Right x -> r === x
  where
    r = merkleRoot t

prop_encodeMerkleTree :: MerkleTree SHA512t_256 -> Property
prop_encodeMerkleTree t
    = case decodeMerkleTree (encodeMerkleTree @BA.Bytes t) of
        Left e -> error (displayException e)
        Right x -> t === x



bench/Main.hs
{-# LANGUAGE AllowAmbiguousTypes #-}
{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE FlexibleContexts #-}
{-# LANGUAGE FlexibleInstances #-}
{-# LANGUAGE LambdaCase #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE PackageImports #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE TypeApplications #-}
{-# LANGUAGE TypeFamilies #-}
{-# LANGUAGE UndecidableInstances #-}

{-# OPTIONS_GHC -fno-warn-orphans #-}

-- |
-- Module: Main
-- Copyright: Copyright © 2019 Kadena LLC.
-- License: MIT
-- Maintainer: Lars Kuhtz <lars@kadena.io>
-- Stability: experimental
--
-- TODO
--
module Main
( main
) where

import Control.DeepSeq

import Criterion
import Criterion.Main

import "crypton" Crypto.Hash
import qualified "cryptonite" Crypto.Hash as CR

import qualified Data.ByteArray as BA
import qualified Data.ByteString as B
import Data.ByteString.Random.MWC
import qualified Data.HashTree as HT
import Data.Maybe

import GHC.Generics

import Numeric.Natural

import System.Random
import qualified System.Random.MWC as MWC

-- internal modules

import qualified Data.MerkleLog as ML

-- -------------------------------------------------------------------------- --
-- Main

main :: IO ()
main = defaultMain
    [ env globalEnv $ \ ~e -> bgroup "main"
        [ bgroup "create tree"
            [ bgroup "SHA512t_256"
                [ createBench @(ML SHA512t_256) e
                , createBench @(HT CR.SHA512t_256) e
                ]
            , bgroup "SHA256"
                [ createBench @(ML SHA256) e
                , createBench @(HT CR.SHA256) e
                ]
            , bgroup "SHA3_256"
                [ createBench @(ML SHA3_256) e
                , createBench @(HT CR.SHA3_256) e
                ]
            , bgroup "BLAKE2b_256"
                [ createBench @(ML Blake2b_256) e
                ]
            ]
        , bgroup "create inclusion proof"
            [ bgroup "SHA512t_256"
                [ proofBench @(ML SHA512t_256) e
                , proofBench @(HT CR.SHA512t_256) e
                ]
            , bgroup "SHA256"
                [ proofBench @(ML SHA256) e
                , proofBench @(HT CR.SHA256) e
                ]
            , bgroup "SHA3_256"
                [ proofBench @(ML SHA3_256) e
                , proofBench @(HT CR.SHA3_256) e
                ]
            , bgroup "BLAKE2b_256"
                [ proofBench @(ML Blake2b_256) e
                ]
            ]
        , bgroup "verify inclusion proof"
            [ bgroup "SHA512t_256"
                [ verifyBench @(ML SHA512t_256) e
                , verifyBench @(HT CR.SHA512t_256) e
                ]
            , bgroup "SHA256"
                [ verifyBench @(ML SHA256) e
                , verifyBench @(HT CR.SHA256) e
                ]
            , bgroup "SHA3_256"
                [ verifyBench @(ML SHA3_256) e
                , verifyBench @(HT CR.SHA3_256) e
                ]
            , bgroup "BLAKE2b_256"
                [ verifyBench @(ML Blake2b_256) e
                ]
            ]
        ]
    ]

-- -------------------------------------------------------------------------- --
-- Merkle Tree Implementations
-- -------------------------------------------------------------------------- --

-- -------------------------------------------------------------------------- --
-- Global Environment

leafCount :: Int
leafCount = 10000

leafMaxSize :: Int
leafMaxSize = 1000

type GlobalEnv = [B.ByteString]

globalEnv :: IO GlobalEnv
globalEnv = do
    gen <- MWC.create
    traverse (randomGen gen) (randomNats leafCount)
  where

randomNats :: Int -> [Natural]
randomNats i = fmap fromIntegral $ take i $ randomRs @Int (0,leafMaxSize) $ mkStdGen 1

-- -------------------------------------------------------------------------- --
-- Create Benchmark

createBench :: forall a . Impl a => GlobalEnv -> Benchmark
createBench = bench (label @a) . nf (tree @a)

-- -------------------------------------------------------------------------- --
-- Proof Benchmark

type ProofEnv a = (Tree a, B.ByteString, Int)

proofEnv :: forall a . Impl a => GlobalEnv -> IO (ProofEnv a)
proofEnv e = return (tree @a e, e !! 277, 277)

-- | Note that this also includes verification of the proof, because that's the
-- only way we can ensure that the resulting proofs are in normal form.
--
proofBench
    :: forall a
    . Impl a
    => GlobalEnv
    -> Benchmark
proofBench e = env (proofEnv @a e)
    $ bench (label @a) . nf (\(t, ix, i) -> proof @a t ix i)

-- -------------------------------------------------------------------------- --
-- Verify Benchmark

type VerifyEnv a = Proof a

verifyEnv :: forall a . Impl a => GlobalEnv -> IO (VerifyEnv a)
verifyEnv e = return $ proof @a (tree @a e) (e !! 277) 277

verifyBench
    :: forall a
    . Impl a
    => GlobalEnv
    -> Benchmark
verifyBench e = env (verifyEnv @a e) $ bench (label @a) . nf verifyThrow
  where
    verifyThrow p
        | verify @a p = ()
        | otherwise = error "benchmark failure"

-- -------------------------------------------------------------------------- --
-- Merkle Tree Implementations
-- -------------------------------------------------------------------------- --

-- -------------------------------------------------------------------------- --
-- Merkle Tree Implementation Class

class (NFData (Tree a), NFData (Root a), NFData (Proof a)) => Impl a where
    type Tree a
    type Proof a
    type Root a

    label :: String
    tree :: [B.ByteString] -> Tree a
    root :: Tree a -> Root a
    proof :: Tree a -> B.ByteString -> Int -> Proof a
    verify :: Proof a -> Bool

-- -------------------------------------------------------------------------- --
-- merkle-log

data MLProof a = MLProof
    {-# UNPACK #-} !(ML.MerkleProof a)
    {-# UNPACK #-} !(ML.MerkleRoot a)
        -- ^ Root of the Tree
    deriving (Generic)

instance NFData (MLProof a)

data ML a

instance HashAlgorithm a => Impl (ML a) where
    type Tree (ML a) = ML.MerkleTree a
    type Proof (ML a) = MLProof a
    type Root (ML a) = ML.MerkleRoot a

    label = "merkle-log"
    tree = ML.merkleTree @a . fmap ML.InputNode
    root = ML.merkleRoot
    proof t ix i =
        let p = case ML.merkleProof (ML.InputNode ix) i t of
                Right x -> x
                Left e -> error $ "proof creation failed in benchmark: " <> show e
        in MLProof p (ML.merkleRoot t)
    verify (MLProof p r) = ML.runMerkleProof p == r

    {-# INLINE label #-}
    {-# INLINE tree #-}
    {-# INLINE root #-}
    {-# INLINE proof #-}
    {-# INLINE verify #-}

-- -------------------------------------------------------------------------- --
-- hash-tree package

data HTProof a = HTProof
    {-# UNPACK #-} !(HT.InclusionProof a)
    {-# UNPACK #-} !B.ByteString
        -- ^ Proof subject (leaf)
    {-# UNPACK #-} !(CR.Digest a)
        -- ^ Root of the Tree
    deriving (Generic)

instance NFData (HTProof a)

instance NFData (HT.MerkleHashTrees B.ByteString a) where
    rnf t = rnf $ HT.digest (HT.size t) t
    {-# INLINE rnf #-}

instance NFData (HT.InclusionProof a) where
    rnf p = rnf (HT.leafIndex p)
        `seq` rnf (HT.treeSize p)
        `seq` rnf (HT.inclusion p)
    {-# INLINE rnf #-}

data HT a

htSettings :: forall a . CR.HashAlgorithm a => HT.Settings B.ByteString a
htSettings = HT.defaultSettings
    { HT.hash0 = CR.hash @B.ByteString @a mempty
    , HT.hash1 = \x -> CR.hash @_ @a (B.singleton 0x00 `B.append` x)
    , HT.hash2 = \x y -> CR.hash @_ @a $ B.concat [B.singleton 0x01, BA.convert x, BA.convert y]
    }

instance (CR.HashAlgorithm a) => Impl (HT a) where
    type Tree (HT a) = HT.MerkleHashTrees B.ByteString a
    type Proof (HT a) = HTProof a
    type Root (HT a) = CR.Digest a

    label = "hash-tree"
    tree = HT.fromList htSettings
    root t = fromJust $ HT.digest (HT.size t) t
    proof t ix _ = HTProof
        (fromJust $ HT.generateInclusionProof (HT.hash1 (htSettings @a) ix) (HT.size t) t)
        ix
        (root @(HT a) t)
    verify (HTProof p subj r) = HT.verifyInclusionProof
        (htSettings @a) (HT.hash1 (htSettings @a) subj) r p

    {-# INLINE label #-}
    {-# INLINE tree #-}
    {-# INLINE root #-}
    {-# INLINE proof #-}
    {-# INLINE verify #-}

